Dockerfile

```diff
@@ -1,124 +1,146 @@
-ENV TERM=xterm
-ENV HOSTNAME=bastion
+RUN cd duo_unix && \
+    ./configure --with-pam --prefix=/usr && \
+    make
 
-ENV MFA_PROVIDER=duo
 
-ENV UMASK=0022
+##
+## Google Authenticator PAM module builder image
+##
+FROM builder as google-authenticator-libpam-builder
 
-ENV DUO_IKEY=
-ENV DUO_SKEY=
-ENV DUO_HOST=
-ENV DUO_FAILMODE=secure
-ENV DUO_AUTOPUSH=yes
-ENV DUO_PROMPTS=1
+ARG AUTHENTICATOR_LIBPAM_VERSION=1.05
+RUN git clone --branch ${AUTHENTICATOR_LIBPAM_VERSION} --single-branch https://github.com/google/google-authenticator-libpam
 
-ENV ENFORCER_ENABLED=true
-ENV ENFORCER_CLEAN_HOME_ENABLED=true
+RUN cd google-authenticator-libpam && \
+    ./bootstrap.sh && \
+    ./configure --prefix=/ && \
+    make
 
-ENV SSH_AUDIT_ENABLED=true
 
-# Enable Rate Limiting
-ENV RATE_LIMIT_ENABLED=true
+##
+## OpenSSH Portable builder image
+##
+FROM builder as openssh-portable-builder
 
-# Tolerate 5 consecutive fairues    
-ENV RATE_LIMIT_MAX_FAILURES=5
+ARG OPENSSH_VERSION=V_7_8_P1
+RUN git clone --branch ${OPENSSH_VERSION} --single-branch https://github.com/openssh/openssh-portable
+
+COPY patches/ /patches/
+
+RUN cd openssh-portable && \
+    git checkout ${OPENSSH_VERSION} && \
+    find ../patches/openssh/** -type f -exec patch -p1 -i {} \; && \
+    autoreconf && \
+    ./configure \
+        --prefix=/usr \
+        --sysconfdir=/etc/ssh \
+        --datadir=/usr/share/openssh \
+        --libexecdir=/usr/lib/ssh \
+        --mandir=/usr/share/man \
+        --with-pid-dir=/run \
+        --with-mantype=man \
+        --with-privsep-path=/var/empty \
+        --with-privsep-user=sshd \
+        --with-md5-passwords \
+        --with-ssl-engine \
+        --disable-wtmp \
+        --with-pam && \
+    make
+
+
+##
+## Bastion image
+##
+FROM builder
+
+LABEL maintainer="erik@cloudposse.com"
 
-# Lock accounts out for 300 seconds (5 minutes) after repeated failures
-ENV RATE_LIMIT_LOCKOUT_TIME=300
-# Sleep N microseconds between failed attempts
-ENV RATE_LIMIT_FAIL_DELAY=3000000
+USER root
 
-#
-# Slack
-#
-ENV SLACK_ENABLED=false
-ENV SLACK_HOOK=sshrc
-ENV SLACK_WEBHOOK_URL=
-ENV SLACK_USERNAME=
-ENV SLACK_TIMEOUT=2
-ENV SLACK_FATAL_ERRORS=true
+## Install sudosh
+ENV SUDOSH_VERSION=0.1.3
+ADD https://github.com/cloudposse/sudosh/releases/download/${SUDOSH_VERSION}/sudosh_linux_386 /usr/bin/sudosh
+RUN chmod 755 /usr/bin/sudosh
 
-#
-# SSH
-#
-ENV SSH_AUTHORIZED_KEYS_COMMAND=none
-ENV SSH_AUTHORIZED_KEYS_COMMAND_USER=nobody
+## Install Duo
+COPY --from=duo-builder duo_unix duo_unix
+RUN (cd duo_unix && make install) && \
+    rm -rf duo_unix
+
+## Install Google Authenticator PAM module
+COPY --from=google-authenticator-libpam-builder google-authenticator-libpam google-authenticator-libpam
+RUN (cd google-authenticator-libpam && make install) && \
+    rm -rf google-authenticator-libpam
+
+## Install OpenSSH Portable
+COPY --from=openssh-portable-builder openssh-portable openssh-portable
+RUN (cd openssh-portable && make install) && \
```

Comment:
What about installing in the builder image to a prefix like `/dist`, then doing a `COPY --from=openssh-portable-builder /dist /`

Reply:
Are the `(....)` necessary? Logically don't seem like it.

Reply:
Agree with you, refactored code to use common approach with `dist` folder